### ISSUE TITLE
Very huge data in production 

### ISSUE THREAD

**Issue by nazmiasri95** [lyst/lightfm#409]
Hi, I'm having problem when I want to make the recommender system in production.
Since the data size is so huge (the dataframe info as below), I will might having problem when I want to serve it using Flask API.

```
<class 'pandas.core.frame.DataFrame'>
Int64Index: 557065 entries, xxx to xxx
Columns: 4932 entries, xx to xxxxxx
dtypes: float64(4932)
memory usage: 20.5 GB
```

In production, I'm planning to retrieve the model, user dictionary and item dictionary. Even if I store interactions matrix in a file, it gonna be very huge file. 

 Is there any better solution or suggestion to create the interactions matrix, to be serve in production ? 

P/S : I already limiting the number of data, and planning to use 600k x 30k matrix later

**Comment by EthanRosenthal** [comment]
It looks like you're storing the data in a dense dataframe. If you avoid storing the data in dense format and solely use sparse matrices, you should use considerably less memory.

For example, let's say that, on average, each of your 600,000 users has interacted with 100 items. Your interactions matrix will consist of 600,000 * 100 floating point numbers which is only 240 MB for 32-bit floats.

**Comment by nazmiasri95** [comment]
Thanks for the response, yesterday I tried to study the code and found out the `interactions` actually can be stored with lesser size, as what @EthanRosenthal mentioned. I avoided storing the data in dense format, instead I store it in key value paired in NoSQL server. 

I needed it to be real time/ live update recommendations, therefore I will store all the clickstream and update users' rating in the nosql, and retrieved it back when user is coming. 

Thanks @SimonCW for your suggestions. I think that's a really brilliant idea ! I should consider to test using ANN (Approximate Nearest Neighbour) as well and see which one is better.
